README.md
[![Travis](https://img.shields.io/travis/heroku/shh.svg)](https://travis-ci.org/heroku/shh)
[![Releases](https://img.shields.io/github/release/heroku/shh.svg)](https://github.com/heroku/shh/releases)
[![GoDoc](https://godoc.org/github.com/heroku/shh?status.svg)](http://godoc.org/github.com/heroku/shh)


System Heuristics Herald (aka Shh)
----

Gathers and relays system metrics

## Install

    go get github.com/heroku/shh

## Configuration

Configuration of shh doesn't use a config file, instead it uses environment variables.

| Environment Var | Type | Explanation | Default |
|:----------------|:-----|:------------|:--------|
| `SHH_INTERVAL` | duration | Polling Interval | 10s |
| `SHH_META` | bool | Report/Collect meta stats | false |
| `SHH_OUTPUTTER` | string | Outputter | stdoutl2metder |
| `SHH_POLLERS` | list of string | Pollers to poll | conntrack,cpu,df,disk,listen,load,mem,nif,ntpdate,processes,self |
| `SHH_SOURCE` | string | Source to emit | |
| `SHH_PREFIX` | string | Metric prefix to use | |
| `SHH_PROFILE_PORT` | string | Profile Port | 0 (off) |
| `SHH_PERCENTAGES` | list of string | Default pollers which should report percentages when applicable | |
| `SHH_DF_TYPES` | list of string | Default DF types | btrfs,ext3,ext4,tmpfs,xfs |
| `SHH_LISTEN` | string | Default network socket info for listen | unix,#shh |
| `SHH_LISTEN_TIMEOUT` | string | Socket timeout duration | `SHH_INTERVAL` |
| `SHH_NIF_DEVICES` | list of string | Devices to poll | eth0 |
| `SHH_NTPDATE_SERVERS` | list of string | NTP Servers | 0.pool.ntp.org,1.pool.ntp.org |
| `SHH_CPU_AGGR` | bool | Whether to only report aggregate CPU usage | true |
| `SHH_LIBRATO_USER` | string | The Librato API User | |
| `SHH_LIBRATO_TOKEN` | string | The Librato API Token | |
| `SHH_LIBRATO_URL` | string | The Librato API User | https://metrics-api.librato.com/v1/metrics |
| `SHH_LIBRATO_BATCH_SIZE` | int | The max number of metrics to submit in a single request | 500 |
| `SHH_LIBRATO_BATCH_TIMEOUT` | duration | The max time metrics will sit un-delivered | `SHH_INTERVAL` |
| `SHH_LIBRATO_ROUND` | bool | Should shh round times to the nearest interval? | true |
| `SHH_NETWORK_TIMEOUT` | duration | Timeout til connect (will retry). And timeout to first header (will assume successful). Used for HTTP(S) endpoints and other network communication | 5s |
| `SHH_CARBON_HOST` | string | Where the Carbon Outputter sends it's data | |
| `SHH_SOCKSTAT_PROTOS` | list of string | Protocols to report sockstats about | TCP,UDP,TCP6,UDP6 |
| `SHH_STATSD_HOST` | string | Where the Statsd Outputter sends it's data | |
| `SHH_STATSD_PROTO` | string | Whether the Stats Outputter uses TCP or UDP | udp |
| `SHH_SYSLOGNG_SOCKET` | string | The location of the syslog-ng socket | /var/lib/syslog-ng/syslog-ng.ctl |
| `SHH_FULL | list of strings | Pollers that should report full metrics. `shh` defaults to minimal | "" |
| `SHH_DISK_FILTER` | regexp | Scan devices that match this regex | (xv|s)d |
| `SHH_PROCESSES_REGEX` | regexp | Scan / extract metrics for processes that match this regex | \A\z |
| `SHH_TICKS` | int | cpu ticks per second: see `getconf CLK_TCK`. Default is probably correct. (temporary until we use cgo) | 100 |
| `SHH_PAGE_SIZE` | int | system page size in bytes: see `getconf PAGESIZE`. Default is probably correct. (temporary until we use cgo) | 4096 |
| `SHH_NAGIOS3_METRIC_NAMES` | list of strings | list of nagios 3 metric names to report stats on, see `nagios3stats -h` | NUMSERVICES,NUMHOSTS,AVGACTSVCLAT,AVGACTHSTLAT,NUMHSTACTCHK5M,NUMSVCACTCHK5M,NUMHSTACTCHK1M,NUMSVCACTCHK1M |
| `SHH_SPLUNK_PEERS_SKIP_VERIFY` | bool | whether or not to skip verification of HTTPS cert on splunk peers endpoint | false |
| `SHH_SPLUNK_PEERS_URL` | string | URL of splunk distributed peers status (e.g. https://user:pass@localhost:8089/services/search/distributed/peers?count=-1 | |
| `SHH_FOLSOM_BASE_URL` | string | URL of exported folsom metrics via folsome\_cowboy or folsom\_webmachine (e.g. https://localhost:5564/) | |
| `SHH_REDIS_URL` | string | URL for Redis as defined by [goredis](https://github.com/xuyu/goredis) (e.g. tcp://auth:password@127.0.0.1:6379/0?timeout=10s&maxidle=1) | tcp://localhost:6379/0?timeout=10s&maxidle=1 |
| `SHH_REDIS_INFO` | string | Description of [INFO](http://redis.io/commands/info): `section0:key0,key1;section1:key0,key1` to pull | clients:connected_clients;memory:used_memory,used_memory_rss;stats:instantaneous_ops_per_sec;keyspace:db0.keys |

For more information on the duration type, see [time.ParseDuration](http://golang.org/pkg/time/#ParseDuration)

The regexp type supports valid regexps documented [here](http://golang.org/pkg/regexp/).

### A note about SHH_OUTPUTTER

The SHH_OUTPUTTER variable *may* not be enough on it's own to get the desired result. For instance, the Librato outputter, requires that `SHH_LIBRATO_USER` and `SHH_LIBRATO_TOKEN` be set.

### A note about SHH_PERCENTAGES

This variable works on "virtual" pollers and computes "percentage used", reporting as "<metric>.perc"

* mem (from the mem poller)
* swap (from the mem poller)
* df (from the df_poller)

## Building Debs

Requirements:

* dpkg (see also `brew install dpkg`)
* go & [gox](https://github.com/mitchellh/gox), which is installed via the Makefile

```bash
make debs
```

Note: You can find debs on the [Github release page](https://github.com/heroku/shh/releases)

## 'Local' Development

1. Obtain a Linux system (only really tested on Ubuntu ATM)
1. Install Go (version 1.4+)
1. Set GOPATH [appropriately](http://golang.org/doc/code.html)
1. `go get github.com/tool/godep`
1. `go get github.com/heroku/shh`
1. cd $GOPATH/src/github.com/heroku/shh
1. go test -v ./...

hack away

## Contributing

The goal for shh is to be a stable, low footprint system metrics
poller, and we welcome contributions, feedback and bug reports to make
that happen.

We're currently focused on supporting GNU/Linux systems, since that's
where we're using shh, but are open to supporting other platforms
provided the low footprint nature is preserved.

Please file bug reports through
[Github Issues](https://github.com/heroku/shh/issues). If you'd like
to contribute changes, please fork and submit a pull request.

## License

Copyright 2013 - 2015, Edward Muller, and contributors

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## 0.9.10 - 2015-03-06

- Added a deb target to the Makefile using gox
- Doc update

## 0.9.9 - 2015-03-06

### Added

- SHH_META: Off by default. This controls meta stats collection and
  reporting. Turn this on to get:

  - Librato stats (number of gauges, number of counters)
  - multi poller's stats
  - listen poller's meta stats

### Removed

- Removed `lo` from the default NIF_DEVICES
- pid.last from the load poller. This is useless on a time series plot.

## 0.9.8 - 2015-01-05

### Changed

- Bugfix: ETS table memory to report in words not bytes.

## 0.9.7 - 2015-01-05

### Changed

- Bugfix: Used memory is a gauge, report it as such.

## 0.9.6 - 2015-01-05

### Changed

- Bugfix: Renamed LinearSliceContainsString to SliceContainsString, removing
    the old implementation, so that we only match on exact matches

## 0.9.5 - 2015-01-05

### Added

- Redis poller.

### Changed

- Update Go Version to 1.4.
- Travis CI Updates

## 0.9.4 - 2014-12-16

### Changed

- Folsom Poller (folsom) now polls for ETS table size and memory.

## 0.9.3 - 2014-11-25

### Changed

- Folsom Poller (folsom) now uses native `folsom_cowboy` API calls to determine
  metric types.

## 0.9.2 - 2014-11-12

### Added

- Folsom Poller (folsom), which adds the ability to fetch metrics from a
  runinning Erlang system. See the [docs](POLLERS.md) for more info.

## 0.9.1 - 2014-11-11

- Retry all low-level delivery errors

### Changed (Breaking)

- Fixed a typo in the per-processes rss metric. byts -> bytes

## 0.9.0 - 2014-11-07

### Changed (Breaking)

- MetricsNameNormalizer was changed so some metrics names may change.
```
    Old # -> _
        - -> _
    New # -> .
        _ -> -
```

## 0.8.9 - 2014-11-06

- Actually catch io.EOFs so we can re-try them. Experience has shown they are
    most likely dropped connections.

## 0.8.8 - 2014-11-06

### Added

- Splunk Search Peers poller (splunksearchpeers), which adds the
  ability to monitor Splunk search clusters. See the
  [docs](POLLERS.md) for more info.

### Changed (Breaking)

- Removed LibratoNetworkTimeout, use NetworkTimeout
  (`SHH_NETWORK_TIMEOUT`) instead.
- Merged listen poller metrics into 3 (all parse errors are treated
  the same) and added prefix `meta` to them all.

### Changed (Non-breaking)

- Refactored listen poller stats counters to avoid mutex
- Removed a bunch of useless logging
- handleListenConnection now part of Listen poller instead of bare.


## 0.8.7 - 2014-11-03

- nagios2stats bugfix

## 0.8.6 - 2014-11-03

- `nagios3stats` poller

## 0.8.5 - 2014-10-31

- Fix reporting of per processes sys / user cpu

## 0.8.4 - 2014-10-30

- Ignore processes w/o names, likely due to a process exiting between
    enumerating the directory entries and reading /proc/<pid>/stat

## 0.8.3 - 2014-10-30

- Generate additional process stats for processes that match
    `SHH_PROCESSES_REGEX`.
- `SHH_PROCESSES_REGEX`: \A\z - Regex of process names to poll and extra
    additional measurements for
- `SHH_TICKS`: 100 - cpu ticks per second. Default should be correct
    for most systems. see `getconf CLK_TCK`. Temporary until we use
    cgo to get it
- `SHH_PAGE_SIZE`: 4096 - kernel page size. Default should be correct
    for most systems. See `getconf PAGESIZE`. Temporary until we use
    cgo to get it.

## 0.8.2 - 2014-10-28

- `SHH_LIBRATO_ROUND`: true - round measurement times to nearest
    interval during submission

## 0.8.1 - 2014-10-24

### Changed (Breaking)

- Some disk poller stats were incorrectly being reported as Gauges.

## 0.8.0 - 2014-10-22

### Added

- `SHH_DF_LOOP` introduced to avoid loopback mounts from showing up in df
    poller (default false)
- `SHH_FULL` introduced to add back full set of metrics from some pollers

### Changed

- Remove `DEFAULT_SELF_POLLER_MODE` in favor of SHH_FULL="self"
- `SHH_CPU_AGGR` defaults to true, eliminating cpu metrics for all cores by
    default
- `SHH_DF_TYPES` removes tmpfs by default
- Default pollers to minimal set. Utilize `SHH_FULL` to get full set of metrics

## 0.7.0 - 2014-10-22

- Update some defaults: `SHH_INTERVAL=60s` & `SHH_LIBRATO_BATCH_TIMEOUT=10s`

## 0.6.4 - 2014-10-22

- Bugfix

## 0.6.3 - 2014-10-22

- SHH_LIBRATO_BATCH_SIZE defaults to 500
- SHH_LIBRATO_NETWORK_TIMEOUT defaults to 5s
- SHH_LIBRATO_BATCH_TIMEOUT defaults to SHH_INTERVAL
- Librato Outputter timeout doesn't start until there is a measurement

## 0.6.2 - 2014-10-22

- Report to librato how many guages / counters are being reported in a batch

## 0.6.1 - 2014-10-22

- `$SHH_LISTEN_TIMEOUT` now controls the timeout on the socket.
- Librato outlet now reports a `User-Agent` header at the request of Librato.
- Timeout errors to the librato api are now reported.
- Better handling of ntpdate sub process error messages.

## 0.6.0 - 2014-10-20

- shh-value cli tool for interacting with the unix socket.
- Latest version of Go (1.3.3) used.
- use github.com/heroku/slog for structured logging (extracted from shh
    originally).
- LISTEN Poller documentation.
- Improved Listen Poller with support for types and units
- Use of Go's logger (over fmt.Println)

## 0.5.0 - 2014-10-09

### Added

- Units. Measurements now have units which the Librato outputter takes
    advantage of. No other outputters currently take advantage of this.

### Changed (Breaking)

- sockstat poller now uses lowercase protocol names in emitted metrics.
    Previously, it broke convention and used uppercase. (i.e. 'UDP' is now
    'udp')
- The percentage calculations for the "mem" and "df" pollers resulted in values
    between 0 and 1. CPU percentages are between 0-100. They now all use 0-100
    instead of 0-1.

## Prior Versions

Sadly, we didn't keep a proper changelog for previous versions. :(


config.go
package shh

import (
	"fmt"
	"net/url"
	"regexp"
	"runtime"
	"time"
)

const (
	DEFAULT_EMPTY_STRING             = ""
	DEFAULT_INTERVAL                 = "60s"                                                              // Default tick interval for pollers
	DEFAULT_OUTPUTTER                = "stdoutl2metder"                                                   // Default outputter
	DEFAULT_POLLERS                  = "conntrack,cpu,df,disk,listen,load,mem,nif,ntpdate,processes,self" // Default pollers
	DEFAULT_PROFILE_PORT             = "0"                                                                // Default profile port, 0 disables
	DEFAULT_DF_TYPES                 = "btrfs,ext3,ext4,xfs"                                              // Default fs types to report df for
	DEFAULT_DF_LOOP                  = false                                                              // Default to not reporting df metrics for loop back filesystems
	DEFAULT_NIF_DEVICES              = "eth0"                                                             // Default interfaces to report stats for
	DEFAULT_NTPDATE_SERVERS          = "0.pool.ntp.org,1.pool.ntp.org"                                    // Default to the pool.ntp.org servers
	DEFAULT_CPU_AGGR                 = true                                                               // Default whether to only report aggregate CPU
	DEFAULT_SYSLOGNG_SOCKET          = "/var/lib/syslog-ng/syslog-ng.ctl"                                 // Default location of the syslog-ng socket
	DEFAULT_SELF_POLLER_MODE         = "minimal"                                                          // Default to only minimal set of self metrics
	DEFAULT_SOCKSTAT_PROTOS          = "TCP,UDP,TCP6,UDP6"                                                // Default protocols to report sockstats on
	DEFAULT_PERCENTAGES              = ""                                                                 // Default pollers where publishing perc metrics is allowed
	DEFAULT_FULL                     = ""                                                                 // Default list of pollers who should report full metrycs
	DEFAULT_LIBRATO_URL              = "https://metrics-api.librato.com/v1/metrics"                       // Default librato url to submit metrics to
	DEFAULT_LIBRATO_BATCH_SIZE       = 500                                                                // Default submission count
	DEFAULT_LIBRATO_BATCH_TIMEOUT    = "10s"                                                              // Default submission after
	DEFAULT_LIBRATO_ROUND            = true                                                               // Round measure_time to interval
	DEFAULT_LISTEN_ADDR              = "unix,#shh"                                                        // listen on UDS #shh
	DEFAULT_DISK_FILTER              = "(xv|s)d"                                                          // xvd* and sd* by default
	DEFAULT_PROCESSES_REGEX          = `\A\z`                                                             // Regex of processes to pull additional stats about
	DEFAULT_TICKS                    = 100                                                                // Default number of clock ticks per second (see _SC_CLK_TCK)
	DEFAULT_PAGE_SIZE                = 4096                                                               // Default system page size (see getconf PAGESIZE)
	DEFAULT_NAGIOS3_METRIC_NAMES     = "NUMSERVICES,NUMHOSTS,AVGACTSVCLAT,AVGACTHSTLAT,NUMHSTACTCHK5M,NUMSVCACTCHK5M,NUMHSTACTCHK1M,NUMSVCACTCHK1M"
	DEFAULT_SPLUNK_PEERS_SKIP_VERIFY = false
	DEFAULT_NETWORK_TIMEOUT          = "5s"
	DEFAULT_REDIS_INFO               = "clients:connected_clients;memory:used_memory,used_memory_rss;stats:instantaneous_ops_per_sec;keyspace:db0.keys" // semi colon seperated section:keya,keyb list
	DEFAULT_REDIS_URL                = "tcp://localhost:6379/0?timeout=10s&maxidle=1"
	DEFAULT_META                     = false
)

var (
	start   = time.Now()
	version string
)

type Config struct {
	version               string
	Interval              time.Duration
	Outputter             string
	Pollers               []string
	Source                string
	Prefix                string
	ProfilePort           string
	Percentages           []string
	Full                  []string
	DfTypes               []string
	DfLoop                bool
	Listen                string
	ListenTimeout         time.Duration
	NifDevices            []string
	NtpdateServers        []string
	CpuOnlyAggregate      bool
	LibratoUrl            *url.URL
	LibratoUser           string
	LibratoToken          string
	LibratoBatchSize      int
	LibratoBatchTimeout   time.Duration
	LibratoRound          bool
	NetworkTimeout        time.Duration
	CarbonHost            string
	SockStatProtos        []string
	StatsdHost            string
	StatsdProto           string
	SyslogngSocket        string
	Start                 time.Time
	DiskFilter            *regexp.Regexp
	UserAgent             string
	ProcessesRegex        *regexp.Regexp
	Ticks                 int
	PageSize              int
	Nagios3MetricNames    []string
	SplunkPeersSkipVerify bool
	SplunkPeersUrl        *url.URL
	FolsomBaseUrl         *url.URL
	RedisUrl              *url.URL
	RedisInfo             string
	Meta                  bool
}

func GetConfig() (config Config) {
	config.Interval = GetEnvWithDefaultDuration("SHH_INTERVAL", DEFAULT_INTERVAL)                                          // Polling Interval
	config.Outputter = GetEnvWithDefault("SHH_OUTPUTTER", DEFAULT_OUTPUTTER)                                               // Outputter
	config.Pollers = GetEnvWithDefaultStrings("SHH_POLLERS", DEFAULT_POLLERS)                                              // Pollers to poll
	config.Source = GetEnvWithDefault("SHH_SOURCE", DEFAULT_EMPTY_STRING)                                                  // Source to emit
	config.Prefix = GetEnvWithDefault("SHH_PREFIX", DEFAULT_EMPTY_STRING)                                                  // Metric prefix to use
	config.ProfilePort = GetEnvWithDefault("SHH_PROFILE_PORT", DEFAULT_PROFILE_PORT)                                       // Profile Port
	config.Percentages = GetEnvWithDefaultStrings("SHH_PERCENTAGES", DEFAULT_PERCENTAGES)                                  // Use Percentages for these pollers
	config.Full = GetEnvWithDefaultStrings("SHH_FULL", DEFAULT_FULL)                                                       // Report full measurements for these pollers
	config.DfTypes = GetEnvWithDefaultStrings("SHH_DF_TYPES", DEFAULT_DF_TYPES)                                            // Default DF types
	config.DfLoop = GetEnvWithDefaultBool("SHH_DF_LOOP", DEFAULT_DF_LOOP)                                                  // Report df metrics for loop back filesystmes or not
	config.Listen = GetEnvWithDefault("SHH_LISTEN", DEFAULT_LISTEN_ADDR)                                                   // Default network socket info for listen
	config.ListenTimeout = GetEnvWithDefaultDuration("SHH_LISTEN_TIMEOUT", config.Interval.String())                       // Listen Poller Socket Timeout
	config.NifDevices = GetEnvWithDefaultStrings("SHH_NIF_DEVICES", DEFAULT_NIF_DEVICES)                                   // Devices to poll
	config.NtpdateServers = GetEnvWithDefaultStrings("SHH_NTPDATE_SERVERS", DEFAULT_NTPDATE_SERVERS)                       // NTP Servers
	config.CpuOnlyAggregate = GetEnvWithDefaultBool("SHH_CPU_AGGR", DEFAULT_CPU_AGGR)                                      // Whether to only report aggregate CPU usage
	config.LibratoUrl = GetEnvWithDefaultURL("SHH_LIBRATO_URL", DEFAULT_LIBRATO_URL)                                       // The Librato API End-Point
	config.LibratoUser = GetEnvWithDefault("SHH_LIBRATO_USER", DEFAULT_EMPTY_STRING)                                       // The Librato API User
	config.LibratoToken = GetEnvWithDefault("SHH_LIBRATO_TOKEN", DEFAULT_EMPTY_STRING)                                     // The Librato API TOken
	config.LibratoBatchSize = GetEnvWithDefaultInt("SHH_LIBRATO_BATCH_SIZE", DEFAULT_LIBRATO_BATCH_SIZE)                   // The max number of metrics to submit in a single request
	config.LibratoBatchTimeout = GetEnvWithDefaultDuration("SHH_LIBRATO_BATCH_TIMEOUT", DEFAULT_LIBRATO_BATCH_TIMEOUT)     // The max time metrics will sit un-delivered
	config.LibratoRound = GetEnvWithDefaultBool("SHH_LIBRATO_ROUND", DEFAULT_LIBRATO_ROUND)                                // Should we round measurement times to the nearest Interval when submitting to Librato
	config.CarbonHost = GetEnvWithDefault("SHH_CARBON_HOST", DEFAULT_EMPTY_STRING)                                         // Where the Carbon Outputter sends it's data
	config.SockStatProtos = GetEnvWithDefaultStrings("SHH_SOCKSTAT_PROTOS", DEFAULT_SOCKSTAT_PROTOS)                       // Protocols to report sockstats about
	config.StatsdHost = GetEnvWithDefault("SHH_STATSD_HOST", DEFAULT_EMPTY_STRING)                                         // Where the Statsd Outputter sends it's data
	config.StatsdProto = GetEnvWithDefault("SHH_STATSD_PROTO", "udp")                                                      // Whether the Stats Outputter uses TCP or UDP
	config.SyslogngSocket = GetEnvWithDefault("SHH_SYSLOGNG_SOCKET", DEFAULT_SYSLOGNG_SOCKET)                              // The location of the syslog-ng socket
	config.ProcessesRegex = GetEnvWithDefaultRegexp("SHH_PROCESSES_REGEX", DEFAULT_PROCESSES_REGEX)                        // The regex to match process names against for collecting additional measurements
	config.Ticks = GetEnvWithDefaultInt("SHH_TICKS", DEFAULT_TICKS)                                                        // Number of ticks per CPU cycle. It's normally 100, but you can check with `getconf CLK_TCK`
	config.PageSize = GetEnvWithDefaultInt("SHH_PAGE_SIZE", DEFAULT_PAGE_SIZE)                                             // System Page Size. It's usually 4096, but you can check with `getconf PAGESIZE`
	config.Nagios3MetricNames = GetEnvWithDefaultStrings("SHH_NAGIOS3_METRIC_NAMES", DEFAULT_NAGIOS3_METRIC_NAMES)         // Which nagios3stats metrics names should we poll
	config.SplunkPeersSkipVerify = GetEnvWithDefaultBool("SHH_SPLUNK_PEERS_SKIP_VERIFY", DEFAULT_SPLUNK_PEERS_SKIP_VERIFY) // If SHH_SPLUNK_PEERS_URL is https, do we need to skip verification?
	config.SplunkPeersUrl = GetEnvWithDefaultURL("SHH_SPLUNK_PEERS_URL", DEFAULT_EMPTY_STRING)                             // URL of splunk search peers api endpoint. Ex: https://user:pass@localhost:8089/services/search/distributed/peers?count=-1
	config.FolsomBaseUrl = GetEnvWithDefaultURL("SHH_FOLSOM_BASE_URL", DEFAULT_EMPTY_STRING)                               // URL of splunk search peers api endpoint. Ex: https://user:pass@localhost:8089/services/search/distributed/peers?count=-1
	config.RedisUrl = GetEnvWithDefaultURL("SHH_REDIS_URL", DEFAULT_REDIS_URL)                                             // URL of redis endpoint. Ex: tcp://auth:pass@localhost:6379/0?timeout=10s&maxidle=1"
	config.RedisInfo = GetEnvWithDefault("SHH_REDIS_INFO", DEFAULT_REDIS_INFO)                                             // section:key1,key2;section2:key1,key2
	config.NetworkTimeout = GetEnvWithDefaultDuration("NETWORK_TIMEOUT", DEFAULT_NETWORK_TIMEOUT)                          // The maximum time to wait for network requests to respond (for both dial and first header when applicable)
	config.Meta = GetEnvWithDefaultBool("SHH_META", DEFAULT_META)                                                          // Should report meta measurements, such as batch sizes for outputters, etc.

	tmp := GetEnvWithDefault("SHH_DISK_FILTER", DEFAULT_DISK_FILTER)
	config.DiskFilter = regexp.MustCompile(tmp)
	config.UserAgent = fmt.Sprintf("shh/%s (%s; %s; %s; %s)", version, runtime.Version(), runtime.GOOS, runtime.GOARCH, runtime.Compiler)
	config.Start = start // Start time
	return
}

func Version() string {
	return version
}
